### PR TITLE
Diagonal transition path

### PR DESCRIPTION
**Pull request by mykola-akulov**

### CODE REVIEW

File: src/aligners/diagonal_transition.rs [hunk 1]
```diff
@@ -115,6 +116,145 @@
...
+            let mut new_cost = s;
+            let mut flag = 0; //This flags shows do we need to save previous vertex
+            match self.gap_variant {
+                GapOpen => {
+                    // If None - we are on the main layer; if layer == Some (idx), then idx - index of the affine layer we are currently are
+                    if let Some(layer_idx) = layer {
+                        let command;
+                        match &self.cm.affine[layer_idx].affine_type {
+                            // I might have mixed up deletion and insertion again :(
+                            InsertLayer => {
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:161]: lgtm

Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:161]:
```suggestion
    let command = match &.. {....};
```

File: src/aligners/diagonal_transition.rs [hunk 2]
```diff
@@ -115,6 +116,145 @@
...
+                                command = CigarOp::AffineInsertion(layer_idx);
+                            }
+                            DeleteLayer => {
+                                new_d = d - 1;
+                                command = CigarOp::AffineDeletion(layer_idx);
+                            }
+                            _ => todo!(),
+                        }
+                        if fronts[s - self.cm.affine[layer_idx].extend as usize]
+                            .affine_mut(layer_idx)[new_d]
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:172]: Replace this by some condition on furthest reaching point in diagonal new_d vs
the current f.r. point

File: src/aligners/diagonal_transition.rs [hunk 3]
```diff
@@ -115,6 +116,145 @@
...
+                        cigar.push(command);
+                    }
+                }
+                GapClose => {
+                    todo!();
+                }
+            }
+
+            if flag == 1 {
+                save((f + new_d) as usize / 2 + 1, (f - new_d) as usize / 2);
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:235]: This is quite ugly :'(

File: src/aligners/diagonal_transition.rs [hunk 4]
```diff
@@ -115,6 +116,145 @@
...
+            } else if flag == 2 {
+                save((f + new_d) as usize / 2, (f - new_d) as usize / 2 + 1);
+            }
+
+            d = new_d;
+            s = new_cost;
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:241]: See if you really need the old values at all. I don't think you do, so you can
just increment `d` and `s` in-place directly.

File: src/aligners/diagonal_transition.rs [hunk 5]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+        //Supports only gap open cost
+        &self,
+        fronts: &Vec<Front<N>>,
+        a: &Sequence,
+        b: &Sequence,
+        p: Pos,
+        layer: Option<usize>,
+        cost: Cost,
+    ) -> (Pos, Option<usize>, CigarOp, Cost) {
+        let diagonal = p.0 as i32 - p.1 as i32;
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:135]: `Fr` instead of `i32`

File: src/aligners/diagonal_transition.rs [hunk 6]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+        cost: Cost,
+        layer: Option<usize>,
+        cigar: &mut Cigar,
+    ) -> (Cost, Fr, Option<usize>) {
+        let s = cost as usize;
+        let mut x;
+        let mut y;
+        if let Some(layer_idx) = layer {
+            let f = fronts[s].affine(layer_idx)[d];
+            x = ((f + d) / 2) as usize;
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:250]: use `fr_to_pos` function

File: src/aligners/diagonal_transition.rs [hunk 7]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+            y = ((f - d) / 2) as usize;
+        } else {
+            let mut f = fronts[s].m()[d];
+            x = ((f + d) / 2 - 1) as usize;
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:254]: why the `-1`?

Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:254]: `x` => `i`
`y` => `j`

File: src/aligners/diagonal_transition.rs [hunk 8]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+            y = ((f - d) / 2 - 1) as usize;
+            //backtracing matches
+            while a[x] == b[y] {
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:257]: `a[i-1] == b[j-1]`

Comment (RagnarGrootKoerkamp) [reply src/aligners/diagonal_transition.rs:257]: Actually, you can reuse the `extend` function in `Backward` direction for this.

File: src/aligners/diagonal_transition.rs [hunk 9]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+                    break;
+                }
+                x -= 1;
+                y -= 1;
+                f -= 1;
+            }
+            x += 1;
+            y += 1;
+        }
+        let parent_pos = self.parent_position(fronts, a, b, Pos(x as u32, y as u32), layer, cost);
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:269]: Unpack the tuple:
`(pos, layer, op, cost) = ...`

File: src/aligners/diagonal_transition.rs [hunk 10]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+    pub fn get_cigar(
+        &self,
+        fronts: &Vec<Front<N>>,
+        a: &Sequence,
+        b: &Sequence,
+        mut cost: Cost,
+    ) -> Cigar {
+        let mut cigar: Cigar = Cigar::default();
+        let mut layer = None;
+        let mut d = a.len() as i32 - b.len() as i32;
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:284]: i32 => Fr

Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:284]: Maybe you want to make a `struct DtState`, containing `d`, `s` and `layer`?

File: src/aligners/diagonal_transition.rs [hunk 11]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+
+    fn parent_state(
+        &self,
+        fronts: &Vec<Front<N>>,
+        a: &Sequence,
+        b: &Sequence,
+        d: Fr,
+        cost: Cost,
+        layer: Option<usize>,
+        cigar: &mut Cigar,
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:243]: Better to just return `CigarOp` I think

File: src/aligners/diagonal_transition.rs [hunk 12]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+            (cost, d, layer) = self.parent_state(fronts, a, b, d, cost, layer, &mut cigar);
+        }
+        cigar.reverse();
+        cigar
+    }
+
+    pub fn get_path(&self, fronts: &Vec<Front<N>>, a: Seq, b: Seq, mut cost: Cost) -> Path {
+        let mut path: Path = vec![];
+        let mut save = |p: Pos| {
+            if let Some(last) = path.last() {
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:295]: make this `if let ... && last != .. { .. }`

File: src/aligners/diagonal_transition.rs [hunk 13]
```diff
@@ -122,6 +122,194 @@ fn fr_to_pos(d: Fr, f: Fr) -> Pos {
...
+                if *last == (p.0 as usize, p.1 as usize) {
+                    return;
+                }
+            }
+            path.push((p.0 as usize, p.1 as usize));
+        };
+        let mut layer = None;
+        let mut p = Pos(a.len() as u32, b.len() as u32);
+        save(p);
+        while p.0 != 0 || p.1 != 0 {
```
Comment (RagnarGrootKoerkamp) [inline src/aligners/diagonal_transition.rs:305]: `|| layer.is_some()`

**Review by RagnarGrootKoerkamp — COMMENTED** [review]
I will move the `diamond` stuff to a separate PR

File: src/aligners/diagonal_transition.rs (hunk 11 above)
Comment (mykola-akulov) [reply src/aligners/diagonal_transition.rs:344]: The problem may be that in the parent_state function I can return more than one CigarOp, and they maybe different types

**Pull request merged**